### ISSUE TITLE
Twinci Token

### ISSUE THREAD

**Issue by duychuongvn** [EthereumCommonwealth/Auditing#499]
# Audit request

... Briefly describe your smart-contract and its main purposes here ...
Twinci Token is a Burnable BEP20  with Fixed supply 5.000.000 tokens, it is a main token for app.twinci.io project
The token is locked in a Vesting contract with 3.000.000 tokens in 365 days where:
 + Developer: receive 25% (750.000 TWIN), every day the system will release 1/365 of 750.000 TWIN 
 + Marketing: receive 40%  (1.200.000 TWIN), every day the system will release 1/365 of 1.200.000  TWIN 
 + Eco system: receive 35 % (1.050.000 TWIN), lock in 365 days and release one time when lock time is over
 
Anyone can call claim methods but the receiver addresses are fixed.

# Source code

1. Token: https://bscscan.com/address/0xAF83F292fcED83032f52ced45ef7DBDdb586441a#code
2. Vesting Contract: https://bscscan.com/address/0x08582ffe7df4400b2dd966fC78BC2bBe81e9b905#code

# Disclosure policy

... Do you want us to publish the report as it is or to notify you privately in case of finding critical mistakes? ...

=> YES

... provide your conditions for publishing the report or leave only standard disclosure policy link ...

[Standard disclosure policy](https://github.com/EthereumCommonwealth/Auditing/blob/master/Standard_disclosure_policy.md).

# Contact information (optional)

... Provide information to contact you or the smart contract-developer in case high severity issues will be found ...

=> developer: caokyhuynh@gmail.com

... Provide information about the media resources of the project you want us to audit (website/ twitter account/ reddit/ telegram channel/ etc.) ...

FB: https://www.facebook.com/twinciio/
Twitter: https://twitter.com/twinciio
Telegram: https://t.me/twinciio_chat
Instagram: https://www.instagram.com/twinci.io/
Discord: https://discord.gg/QmZrtr6VYc
Youtube: https://www.youtube.com/c/Twinci
Medium: https://twinci.medium.com/
Github: https://github.com/twinciio

# Platform

... In which network will your contract be deployed? (EOS/TRX/ETC/ETH/CLO/UBQ/something else ) ...
=> BSC

**Comment by yuriy77k** [comment]
@duychuongvn The audit fee is 759 USDT.
You may send USDT (ERC20 or BEP20) to: 0xb9662e592f2f0412be62f0833ca463a9b1aabebb
or USDT (TRC20) to: TBzUKbek9AYVBwf91ykh3KY4Ushk95SCiB

The estimated auditing time - 3 days after payment.

**Comment by yuriy77k** [comment]
# Twinci Token Security Audit Report

# 1. Summary

[Twinci Token](https://twinci.io/) smart contract security audit report performed by [Callisto Security Audit Department](https://github.com/EthereumCommonwealth/Auditing)

>
> Twinci Token is a Burnable BEP20 with Fixed supply 5.000.000 tokens, it is a main token for app.twinci.io project
> The token is locked in a Vesting contract with 3.000.000 tokens in 365 days where:
> 
> - Developer: receive 25% (750.000 TWIN), every day the system will release 1/365 of 750.000 TWIN
> - Marketing: receive 40% (1.200.000 TWIN), every day the system will release 1/365 of 1.200.000 TWIN
> - Eco system: receive 35 % (1.050.000 TWIN), lock in 365 days and release one time when lock time is over
>
> Anyone can call claim methods but the receiver addresses are fixed.

- Website: https://twinci.io/ 
- FB: https://www.facebook.com/twinciio/
- Twitter: https://twitter.com/twinciio
- Telegram: https://t.me/twinciio_chat
- Instagram: https://www.instagram.com/twinci.io/
- Discord: https://discord.gg/QmZrtr6VYc
- Youtube: https://www.youtube.com/c/Twinci
- Medium: https://twinci.medium.com/
- Github: https://github.com/twinciio

# 2. In scope

1. Token: https://bscscan.com/address/0xAF83F292fcED83032f52ced45ef7DBDdb586441a#code

2. Vesting Contract: https://bscscan.com/address/0x08582ffe7df4400b2dd966fC78BC2bBe81e9b905#code


# 3. Findings

In total, **2 issues** were reported including:

- 1 low severity issues.

- 1 notes.

No critical security issues were found.

## 3.1. Known vulnerabilities of BEP-20 token

### Severity: low

### Description

1. Lack of transaction handling mechanism issue. [WARNING!](https://gist.github.com/Dexaran/ddb3e89fe64bf2e06ed15fbd5679bd20)  This is a very common issue and it already caused millions of dollars losses for lots of token users! More details [here](https://docs.google.com/document/d/1Feh5sP6oQL1-1NHi-X1dbgT3ch2WdhbXRevDN681Jv4/edit).

### Recommendation

Add the following code to the `transfer(_to address, ...)` function:

```
require( _to != address(this) );

```

2. ERC20 is a widely used standard across the Ethereum ecosystem. It is reasonable to assume that ERC20 tokens could be "accidentally" deposited to this contract even though it is not intentional.

Every user on the entire Ethereum ecosystem can send ERC20 tokens to this contract and he will have no ability to extract it back unless there is a special "ERC20-rescue" function implemented in your contract. It is advised to implement this function.

*Example: here is [BAT contract address](https://etherscan.io/address/0x0d8775f648430679a709e98d2b0cb6250d2887ef). As you can see the contract itself holds $497,000 worth of different ERC20 tokens - all these tokens are permanently "stuck" inside the contract and therefore uselessly lost.*

### Recommendation

A simple "ERC20-rescue" function can solve the problem.

```js
interface IERC20 {
  function transfer(address _to, unit _amount);
}

function rescueERC20(address _token, uint256 _amount) external onlyOwner {
    IERC20(_token).transfer(owner(), _amount);
  }
```

## 3.2. Unnecessary gas wasting.

### Severity: note

### Description

In the contract `TwinciVestingContract` the function `erc20safeTransferFrom()` member of `ERC20TransferProxy` contract used for transfer token from contract address to destination address. But in compare with `token.transfer(to)` function it require much more Gas due using cross-contract calls and calling `approve()` function. 

Also, calling the function `erc20safeTransferFrom()` required that caller (`TwinciVestingContract` contract) should has an `Operator` role. That means that `Owner` may remove `Operator` role from `TwinciVestingContract` contract and disallow token transferring from it.


# 4. Conclusion

The audited smart contract can be deployed. Only low severity issues were found during the audit.

**Comment by duychuongvn** [comment]
Hi Yuriy,
Thank you for your email, I am waiting for confirmation from our customer.
They said they want a cheaper and faster service but they have not
confirmed yet.
I saw you  already audit this request, thank you for it, but the customer
may not pay for it because they have not approved the request.
I will work with them for this audit fee.

I really appreciate you and your team for  this service.

Best regards,
Chuong Huynh

Vào Th 2, 10 thg 5, 2021 vào lúc 17:56 Yuriy ***@***.***>
đã viết:
